File: NeetCode/Stack/Largest_Rectangle_in_a_Histogram/largest__rectangle_in_a__histogram.py
from typing import Optional, List


class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        maxArea = 0
        stack = []

        for index, height in enumerate(heights):
            start = index
            while stack and stack[-1][1] > height:
                idx, h = stack.pop()
                maxArea = max(maxArea, h * (index - idx))
                start = idx

            stack.append((start, height))

        for index, height in stack:
            maxArea = max(maxArea, height * (len(heights) - index))

        return maxArea


if __name__ == "__main__":
    sol = Solution()
    heights = [7, 1, 7, 2, 2, 4]

    print(sol.largestRectangleArea(heights))
    print("Running Solution...")

File: NeetCode/Stack/Largest_Rectangle_in_a_Histogram/Largest_Rectangle_in_a_Histogram.cpp
// Largest_Rectangle_in_a_Histogram.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    int largestRectangleArea(vector<int> &heights)
    {
        int maxArea = 0;
        vector<pair<int, int>> stack;
        int height = 0;
        int start = 0;
        int idx = 0;
        int h = 0;
        for (int index = 0; index < heights.size(); index++)
        {
            height = heights[index];
            start = index;

            while (!stack.empty() && stack.back().second > height)
            {
                idx = stack.back().first;
                h = stack.back().second;
                stack.pop_back();
                maxArea = max(maxArea, h * (index - idx));
                start = idx;
            }

            stack.push_back({start, height});
        }

        // if the stack isn't empty after iterating thru, iterate thru the stack

        for (pair<int, int> tmp : stack)
        {
            idx = tmp.first;
            height = tmp.second;
            maxArea = max(maxArea, (int)(height * (heights.size() - idx)));
        }

        return maxArea;
    }
};
